tests/FSharp.Compiler.ComponentTests/Language/CopyAndUpdateTests.fs

```diff
@@ -0,0 +1,371 @@
+            Range = { StartLine = 14
+                      StartColumn = 45
+                      EndLine = 14
+                      EndColumn = 48 }
+            Message = @"This expression was expected to have type
+    'string'    
+but here has type
+    'float'    "
+        }
+    ]
+
+[<Fact>]
+let ``Anonymous record with nested copy-and-update can change shape``() =
+    FSharp """
+module CopyAndUpdateTests
+
+type RecTy = { D: int; E: string option }
+
+let start = {| R = { D = 2; E = Some "e" }; S = 3 |}
+
+let actual = {| start with R.E = None; S = "May I be a string now?"; T = 4 |}
+
+let expected = {| R = { D = 2; E = None }; S = "May I be a string now?"; T = 4 |}
+
+if actual <> expected then
+    failwith "actual does not equal expected"
+    """
+    |> withLangVersionPreview
+    |> compileExeAndRun
+    |> shouldSucceed
+
+[<Fact>]
+let ``Anonymous record in a nominal record with nested copy-and-update cannot change shape``() =
+    FSharp """
+module CopyAndUpdateTests
+
+type RecTy = { D: int; E: {| A: int |} }
+
+let f x = { x with E.A = "May I be a string now?" }
+    """
+    |> withLangVersionPreview
+    |> typecheck
+    |> shouldFail
+    |> withResult {
+        Error = Error 1
+        Range = { StartLine = 6
+                  StartColumn = 26
+                  EndLine = 6
+                  EndColumn = 50 }
+        Message = "This expression was expected to have type
+    'int'    
+but here has type
+    'string'    "
+    }
+
+[<Fact>]
+let ``Nested copy-and-update does not compile when referencing invalid fields``() =
+    FSharp """
+module CopyAndUpdateTests
+
+type NestdRecTy = { B: string; G: {| a: int |} }
+
+type RecTy = { D: NestdRecTy; E: string option }
+
+let t1 x = { x with D.B.A = "a" }
+let t2 x = { x with D.C = "a" }
+let t3 x = { x with D.G.b = "a" }
+let t4 x = { x with C.D = "a" }
+let t5 (x: {| a: int; b: NestdRecTy |}) = {| x with b.C = "a" |}
+let t6 (x: {| a: int; b: NestdRecTy |}) = {| x with b.G.b = "a" |}
+let t7 (x: {| a: int; b: NestdRecTy |}) = {| x with c.D = "a" |}
+    """
+    |> withLangVersionPreview
+    |> typecheck
+    |> shouldFail
+    |> withDiagnostics [
+        (Error 39, Line 8, Col 25, Line 8, Col 26, "The record label 'A' is not defined.")
+        (Error 1129, Line 9, Col 23, Line 9, Col 24, "The record type 'NestdRecTy' does not contain a label 'C'.")
+        (Error 1129, Line 10, Col 25, Line 10, Col 26, "The record type '{| a: int |}' does not contain a label 'b'.")
+        (Error 39, Line 11, Col 21, Line 11, Col 22, "The namespace or module 'C' is not defined.")
+        (Error 1129, Line 12, Col 55, Line 12, Col 56, "The record type 'NestdRecTy' does not contain a label 'C'.")
+        (Error 1129, Line 13, Col 57, Line 13, Col 58, "The record type '{| a: int |}' does not contain a label 'b'.")
+        (Error 1129, Line 14, Col 53, Line 14, Col 54, "The record type '{| a: int; b: NestdRecTy |}' does not contain a label 'c'.")
+    ]
```

Comment:
Could you please add test for something like
```fsharp
type A = { B: string }
type Foo = { ``A.B``: string; A: A}
let x = {``A.B`` = "foo"; A = { B = "foo" } }
let y = { x with ``A.B`` = "bar"; A.B = "bar" }
```

Reply:
`A.B = "bar"` doesn't work because modules/types are prioritized over fields when resolving the long identifier. `Foo.A.B = "bar"` will give you the expected result.

Reply:
`let y = { x with ``A.B`` = "bar"; A.B = "bar" }` will complain about using fields from different records. Not sure how easy it would be to implement a special diagnostic here, it really depends on what the users intends to do - nested or regular update.